### PR TITLE
Shuffle end to end v0

### PR DESCRIPTION
**Pull request by dimroc**
## Motivation

First end to end working version of the shuffle CLI, including new, node, account, deploy, console, test.

asciicinema: https://asciinema.org/a/1ysGZcwF3Vy5E4bFPaFuHg70F

### Review Instructions

This PR has a lot of files that were duplicated to facilitate the creation of a Shuffle package, mostly by placing the files in `move/examples`. Many of them were autogenerated by Transaction Builder or copied from stdlib, and should be ignored for review.

Exceptions to this are
1. `shuffle/move/examples/Message/txn_builders/helper.ts`
1. `shuffle/move/examples/Message/txn_builders/mod.ts`

Also be sure to review `shuffle/cli/repl.ts`

CleanShot 2021-10-01 at 09 29 39@2x

### Developer Instructions

1. `cd ~/workspace/diem`
1. `rm -rf ~/tmp/nodiemdiem && RUST_BACKTRACE=1 cargo run -p shuffle -- new ~/tmp/project`
1. `cargo run -p shuffle -- node ~/tmp/project`
1. open new terminal
1. `cargo run -p shuffle -- account ~/tmp/project shuffle/cli/new_account.key`
1. `cargo run -p shuffle -- deploy ~/tmp/project shuffle/cli/new_account.key`
1. `cargo run -p shuffle -- console ~/tmp/project` # Then `await Shuffle.resources()`
1. `cargo run -p shuffle -- test ~/tmp/project`

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A diem-sdk 0.0.2 (workspace, path 'sdk')
    * features: client, default, diem-client
  A hyper-tls 0.5.0 (transitive third-party, crates.io)
    * features: [none]
  A tokio-native-tls 0.3.0 (transitive third-party, crates.io)
    * features: [none]
  M diem-client 0.0.2 (workspace, path 'sdk/client')
    * added features: async, default, faucet, futures, reqwest, tokio, tokio-tungstenite, websocket
    * (unchanged features: blocking, ipnet, ureq)
  M reqwest 0.11.2 (direct third-party, crates.io)
    * added features: __tls, default, default-tls, hyper-tls, native-tls-crate, tokio-native-tls
    * (unchanged features: blocking, json, serde_json, stream)
```

File: shuffle/cli/src/deploy.rs [hunk 1]
```diff
@@ -1,79 +1,110 @@
...
-        &[move_stdlib_dir.as_str(), diem_framework_dir.as_str()],
-    )?;
 
-    let publish_txn = new_account.sign_with_transaction_builder(
-        factory.payload(TransactionPayload::Module(Module::new(compiled_module))),
-    );
+    // ================= Send a module transaction ========================
 
-    send(&client, publish_txn)?;
+    let mut new_account = LocalAccount::new(derived_address, new_account_key, 0);
```
Comment (avin1100) [inline shuffle/cli/src/deploy.rs:73]: Does this only work assuming that the new_account has already been made an is in the BC?

Comment (dimroc) [reply shuffle/cli/src/deploy.rs:73]: This command assumes that the new account was already made via `shuffle account`, or it errors out with account not found. Not sure what you mean by BC here?

We're reusing what was there before: 
https://github.com/diem/diem/blob/6d265b6bb7798b5e6b61487adc89b5fc3ec57f60/shuffle/cli/src/deploy.rs#L44-L48

File: shuffle/move/examples/stdlib/Nursery/sources/Compare.move [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/// Utilities for comparing Move values based on their representation in BCS.
+module Std::Compare {
```
Comment (sunmilee) [inline shuffle/move/examples/stdlib/Nursery/sources/Compare.move:2]: Where do we use this?

Comment (sunmilee) [inline shuffle/move/examples/stdlib/Nursery/sources/Compare.move:2]: Why is it in the examples directory?

Comment (dimroc) [reply shuffle/move/examples/stdlib/Nursery/sources/Compare.move:2]: This stdlib copy will be removed in the future. This right now serves as a stop gap to allow a relative reference ("../stdlib") in generated shuffle projects.

File: shuffle/cli/src/deploy.rs [hunk 3]
```diff
@@ -1,79 +1,110 @@
...
+use move_lang::compiled_unit::{CompiledUnit, NamedCompiledModule};
+use move_package::compilation::{
+    compiled_package::{CompiledPackage, OnDiskCompiledPackage},
+    package_layout::CompiledPackageLayout,
 };
+use std::{collections::HashSet, convert::TryFrom, path::Path};
+
+// TODO: Deploys the Message Move Package to the sender's address
+// 0x24163afcc6e33b0a9473852e18327fa9. This has been hardcoded via named
+// addresses in diem_framework_named_addresses.
```
Comment (sunmilee) [inline shuffle/cli/src/deploy.rs:31]: Update comment

File: shuffle/cli/src/console.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+use crate::shared;
+use anyhow::Result;
+use std::{path::Path, process::Command};
+
+pub fn handle(project_path: &Path) -> Result<()> {
+    let _config = shared::read_config(project_path)?;
+
+    // TODO: Fix hardcoding of mod.ts by copying it into project path and referencing
+    // it via project_path.join(...). Remove Message pkg hardcode and iterate over pkgs.
```
Comment (sunmilee) [inline shuffle/cli/src/console.rs:12]: Update comment to include TODO to pass in project name

Also probably best to rename Message to Project

Comment (dimroc) [reply shuffle/cli/src/console.rs]: Removed TODO and renamed package to `main/` to mirror rust.

File: shuffle/cli/src/console.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    let deno_bootstrap = format!(
+        r#"import * as Shuffle from "{}/repl.ts";
+        import * as TxnBuilder from "{}/Message/txn_builders/mod.ts";
```
Comment (sunmilee) [inline shuffle/cli/src/console.rs]: How about "Message" are we leaving this hardcoded as message? or "project"?

**Review by sunmilee — DISMISSED** [review]

**Pull request closed without merging**